### DISCUSSION TITLE
Details panel on default startup and shortcut keyboard

### DISCUSSION THREAD

**Discussion by giorgiobrit** [TT-ReBORN/Georgia-ReBORN#111] [Q&A] [answered]
I wanted to ask if there is a way to set the details panel by default when starting georgia reborn. Also if it is possible via keyboard shortcut to switch from biography/playlist panel to details panel. Thank you

**Comment by TT-ReBORN** [comment, marked as answer]
You can deactivate in top menu Options > Playlist > Show playlist on startup, then the theme will always start with the Details panel.

If you want a custom shortcut to the Details panel, you would need to do it yourself because jumpSearch and other shortcuts are already used and I won't add this support.

If you want to do it yourself, open gr-callbacks.js and add these for example ( CTRL + D shortcut ):

In `function on_char(code) {`
Replace original code:
```
// Switch back to Playlist
if (pref.layout === 'default' && !displayPlaylist && !displayLibrary) {
	btns.details.onClick();
}
```

with:
```
// Switch back to Playlist
if (pref.layout === 'default' && !displayPlaylist && !displayLibrary && !utils.IsKeyPressed(VK_CONTROL)) { // Exclude CTRL key
	btns.details.onClick();
}
```

And add this code at the end in `function on_key_down(vkey) {`
```
	// * Switch to Details panel on CTRL + D
	if (CtrlKeyPressed && utils.IsKeyPressed(VK_KEY_D)) {
		if (pref.layout !== 'artwork') {
			btns.details.onClick();
			if (displayPlaylist && !pref.lyricsActiveState) pref.displayLyrics = false;
		}
		else if (pref.layout === 'artwork') {
			btns.playlistArtworkLayout.onClick();
			if (displayPlaylistArtworkLayout) pref.displayLyrics = false;
		}
		playlist.on_size(ww, wh);
		resizeArtwork(true);
		initButtonState();
		window.Repaint();
	}
```

-TT

**Reply by giorgiobrit** [reply]
Big big thank you!!! thanks for all awesome work!

**Comment by giorgiobrit** [comment]
Hello TT and thank you very much for hard work on Georgia Reborn, the last beta is great! I wanted to ask if this modification to have a keyboard shortcut is still valid? I tried it as you described, but unfortunately, everything crashes. Since I control Georgia Reborn using a gamepad that emulates the keyboard, it would be great if there could be a shortcut between the details screen and the playlist screen, and vice versa. I hope you can help me. Thanks in advance!"

**Reply by TT-ReBORN** [reply]
Hi @giorgiobrit,

yes, the code above is not compatible anymore with the latest DEV version because of the last rewrite.

You now need to use this code here:

Open `gr-callbacks.js` and in `function on_char(code) {`
Replace original code:
```
		// Switch back to Playlist
		if (grSet.layout === 'default' && grm.ui.displayDetails) {
			grm.button.btn.details.onClick();
		}
		else if (grSet.layout === 'artwork' && !grm.ui.displayPlaylistArtwork && !grm.ui.displayLibrary) {
			grm.button.btn.playlist.onClick();
		}
```

with:
```
		// Switch back to Playlist
		if (grSet.layout === 'default' && !grm.ui.displayPlaylist && !grm.ui.displayLibrary && !utils.IsKeyPressed(VKey.CONTROL)) { // Exclude CTRL key
			grm.button.btn.details.onClick();
		}
		else if (grSet.layout === 'artwork' && !grm.ui.displayPlaylistArtwork && !grm.ui.displayLibrary && !utils.IsKeyPressed(VKey.CONTROL)) {
			grm.button.btn.playlist.onClick();
		}
```

Now open `gr-main.js` and go to `handleKeyAction(vkey) {`
Add this code block:
```
			[VKey.KEY_D]: {
				ctrl: () => {
					if (!this.displayDetails && grSet.layout === 'default' || !this.displayPlaylist && grSet.layout === 'artwork') {
						this.displayPanel('details', true);
					} else {
						this.displayPanel('playlist', true);
					}
				}
			},
```

like this:
```
		const keyActions = {
			[VKey.ADD]: {
				ctrlNoShift: () => grm.display.handleDisplayScaleKeyAction('increase'),
				altNoShift: () => grm.display.handleDisplayScaleKeyAction('reset'),
				ctrlShift: () => setRating('+')
			},
			[VKey.SUBTRACT]: {
				ctrlNoShift: () => grm.display.handleDisplayScaleKeyAction('decrease'),
				altNoShift: () => grm.display.handleDisplayScaleKeyAction('reset'),
				ctrlShift: () => setRating('-')
			},
			[VKey.KEY_D]: {
				ctrl: () => {
					if (!this.displayDetails && grSet.layout === 'default' || !this.displayPlaylist && grSet.layout === 'artwork') {
						this.displayPanel('details', true);
					} else {
						this.displayPanel('playlist', true);
					}
				}
			},
			[VKey.KEY_Z]: {
				ctrl: () => fb.RunMainMenuCommand('Edit/Undo')
			}
		};
```

Save and restart foobar, now it should work once again.

-TT

**Reply by giorgiobrit** [reply]
Many many thanks Tom!!! You're always kind! Thank you for everything! Enjoy your holidays.

**Comment by giorgiobrit** [comment]
Hi Tom, I tried to do as you told me, but it seems that something isn't working.

**Reply by TT-ReBORN** [reply]
I see, you want to have it toggleable. I edited and updated the new code here:
https://github.com/TT-ReBORN/Georgia-ReBORN/discussions/111#discussioncomment-10570295

Try it now, it should work...

**Comment by giorgiobrit** [comment]
Hi tom, now works very well but If I start Foobar and then don't right-click inside the interface, this keyboard shortcut doesn't work. Once I click the mouse button on any part of the interface, the shortcut works. I'm not sure if I explained myself clearly. It seems that to activate this shortcut, it's necessary to press the mouse button. Do you know why this happens?

**Reply by TT-ReBORN** [reply]
Hi @giorgiobrit,

I have taken a look at your described issue and it is a Spider Monkey Panel problem,  this was already reported by @regorxxx here https://github.com/TheQwertiest/foo_spider_monkey_panel/issues/195 

It is a panel focus issue and I tried some workarounds but unfortunately they do not work... 

So you still need to press the left or right mouse button to get the panel focus in order to make the keyboard shortcuts to work. 

-Tom

**Reply by giorgiobrit** [reply]
Thanks Tom! Let's hope that Spider Monkey gets updated soon for the 64-bit version as well.
